### ISSUE TITLE
FullscreenWrapper doesn't dismiss when isPresented becomes false

### ISSUE THREAD

**Issue by vdka** [SwiftUI-Plus/Presentation#2]
## General Information

 - Project version:
1.1.0
 
 - Platform/OS version: 
iOS 15 (only one tested)
 
 - IDE version:
Xcode 13
 
 - Devices: 
iPad Pro

## Describe the bug
When the following code is used, the Dismiss button fails to dismiss the fullscreen view. 

```swift
.present(isPresented: $showSuccess, transition: .crossDissolve, style: .overFullScreen, onDismiss: returnToOverview) {
    SuccessOverlay(...) {
        // Included within SuccessOverlay
        Button("Dismiss", action: { showSuccess = false })
    }
}
```

## Expected behaviour
`FullscreenWrapper` should be dismissed and any onDismiss should be called

I managed to make it work somewhat by making the following changes, but I am sure I am misunderstanding something around how `FullscreenWrapper` works. I am with this change seeing the following error in the console when dismissing.
> Unbalanced calls to begin/end appearance transitions for < ... >.
```
diff --git a/Sources/Presentation/Fullscreen.swift b/Sources/Presentation/Fullscreen.swift
index b750c27..1e5d44d 100644
--- a/Sources/Presentation/Fullscreen.swift
+++ b/Sources/Presentation/Fullscreen.swift
@@ -30,6 +30,8 @@ internal struct FullscreenView<Content: View>: UIViewControllerRepresentable {
 
 private final class FullscreenWrapper<Content: View>: UIViewController {
 
+    var controller: UIViewController?
+
     var isPresented: Binding<Bool>
     var isModal: Bool = false
     var transition: UIModalTransitionStyle = .coverVertical
@@ -64,6 +66,10 @@ private final class FullscreenWrapper<Content: View>: UIViewController {
                     self?.onDismiss?()
                 }
                 present(controller, animated: true)
+                self.controller = controller
+            } else if !isPresented.wrappedValue && isAlreadyPresented {
+                self.controller?.dismiss(animated: true)
             }
         }
     }

```

Note this patch is applied on top of the PR #1

**Comment by shaps80** [comment]
Great catch I will review the PR – thank you for your contribution 👍

**Comment by shaps80** [comment]
The following does _not_ dismiss when `showPresentation` becomes `false` which I agree, is an issue. But This still fails with your code AFAICT.

```swift
import SwiftUI
import Presentation

struct ContentView: View {
    @State private var showPresentation: Bool = false

    var body: some View {
        Button {
            showPresentation = true
        } label: {
            Text("Show Presentation")
        }
        .present(isPresented: $showPresentation, style: .overFullScreen) {
            print("Dismissed")
        } content: {
            GeometryReader { geo in
                ZStack {
                    Color(.systemBackground)

                    Button {
                        showPresentation = false
                    } label: {
                        Text("Dismiss")
                    }
                }
            }
        }

    }
}

struct ContentView_Previews: PreviewProvider {
    static var previews: some View {
        ContentView()
    }
}
```

**Comment by shaps80** [comment]
Interestingly I also notice (maybe specific to iPad) that the fullscreen present/dismiss transitions are terrible. I think this definitely needs more work in general.

**Comment by vdka** [comment]
Ok, I think I've introduced a bit of confusion here by opening the issue at the same time, my view is that the Issue is a separate to what I am fixing in the #1.  Definitely hold off on the PR for now though these seem related.

I first noticed that when I dismissed by using `@Environment(\.presentation).wrappedValue.dismiss()` within the presented view that the `onDismiss` wasn't being called, then after I went and fixed that I noticed while presenting a view with a Button action that did `isPresented = false` (from the presenting screen so can't access the presented views `@Environment(\.presentation)`) that the `FullscreenWrapper` didn't dismiss.

I think there are 2 code paths that can dismiss the ViewController presented on `FullscreenWrapper` 
1. The presenting view controller could change the value passed in `isPresented: Binding<Bool>`, 
2. The presented view itself could use `@Environment(\.presentation)` to dismiss itself
Both of these need to call `onDismiss` I believe that the patch I posted only called `onDismiss` when dismissed by `@Environment(\.presentation)`

I think the following change is what's needed to call `onDismiss` in both cases, it seems like there is 2 pieces of state that need to stay in sync here, maybe this could be written some way to only have the one binding trigger dismiss and present.

```swift
        if !isAlreadyPresented && isPresented.wrappedValue {
            let controller = viewController(rootView: content(), isModal: isModal, transition: transition, style: style, onDismiss: { [weak self] in
                self?.isPresented.wrappedValue = false
                self?.onDismiss?()
            })
            present(controller, animated: true)
            self.controller = controller
        } else if isAlreadyPresented && !isPresented.wrappedValue {
            if let controller = controller, !controller.isBeingDismissed {
                self.controller?.dismiss(animated: true)
                self.onDismiss?()
            }
        }
    }
```

**Comment by vdka** [comment]
I've expanded on the example you've provided for testing, this let's you test that onDismiss is called for both `@Environment(\.presentation).wrappedValue.dismiss()` and for `isPresented = false`
```swift
import SwiftUI
import Presentation

struct ContentView: View {
    @State private var showPresentation: Bool = false
    @State var color = Color(red: .random(in: 0...1), green: .random(in: 0...1), blue: .random(in: 0...1))
    @State var number = Int.random(in: -100...100)

    var body: some View {
        VStack {
            Text("onDismiss will randomize these 2 things").font(.caption).foregroundColor(.secondary)
            HStack {
                Text("Random number is \(number)")
                Rectangle().foregroundColor(color).frame(width: 20, height: 20)
            }
            Spacer()
            Button("Show Presentation", action: { showPresentation = true })
                .padding()
            Spacer()
        }
        .present(isPresented: $showPresentation, transition: .crossDissolve, style: .overFullScreen, onDismiss: randomizeThings) {
            OverlayView {
                Toggle("showPresentation", isOn: $showPresentation)
            }
        }
        .preferredColorScheme(.dark)
    }
    
    func randomizeThings() {
        print(#function)
        number = Int.random(in: -100...100)
        color = Color(red: .random(in: 0...1), green: .random(in: 0...1), blue: .random(in: 0...1))
    }

    struct OverlayView<Content: View>: View {

        @Environment(\.presentation) var presentation

        let content: () -> Content

        init(@ViewBuilder content: @escaping () -> Content) {
            self.content = content
        }

        var body: some View {
            VStack {
                content()
            }
                .padding()
                .frame(minWidth: 200, maxWidth: 333, minHeight: 200)
                .background(RoundedRectangle(cornerRadius: 16, style: .continuous).fill(Color(.secondarySystemBackground)))
                .overlay(closeButton, alignment: .topTrailing)
                .padding()
                .frame(maxWidth: .infinity, maxHeight: .infinity)
                .background(
                    Color.gray.opacity(0.2).onTapGesture(perform: { presentation.wrappedValue.dismiss() })
                )
        }

        var closeButton: some View {
            Button(action: { presentation.wrappedValue.dismiss() }) {
                Image(systemName: "xmark")
                    .font(.system(.body).bold())
                    .foregroundColor(.secondary)
                    .padding()
                    .contentShape(Rectangle())
            }
        }
    }
}
```

**Comment by shaps80** [comment]
Hey sorry, I sent the day on this hahaha so I took a different approach to simplify the approach as I wasn't happy with a few other unrelated issues as well.

What you've mentioned above however does indeed now work as expected 👍

**Comment by vdka** [comment]
Thanks for that! Did some testing yesterday, found no issues.

There is one change I noticed that slipped through I am not sure was intended. The `.backgroundColor` for the `UIHostingController.view` [was](https://github.com/SwiftUI-Plus/Presentation/pull/3/commits/40a29c29aaa5b57c5203b732251d1fb2bddd8528#diff-0c0fe29113442a7bbca93a9b60af42e56c543b31abf18ef88a47e70af2d443afL42) set to `.clear` in  `1.1.0` where now in `1.2.0` it is left as default (something opaque). I worked around this by injecting a `UIView` and traversing the view hierarchy until I found the `UIHostingController` with [Introspect](https://github.com/siteline/SwiftUI-Introspect)
```swift
.introspectViewController { $0.view.backgroundColor = .clear }
```
so it's still fine for my purpose, but perhaps the ability to customize the `UIHostingController` could be a further improvement?
